### PR TITLE
Enable S.R.CS.Unsafe and System.Memory tests for UAP

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/dotnet/corefx/issues/23406

**The Span tests fail for uapaot when we try to access the elements of a Span of a reference type like string using the provided indexer:**
```C#
public static void TestingSpanReferenceTypeForUAPAOT()
{
    ReadOnlySpan<byte> span1 = new byte[] { 1 };
    byte temp1 = span1[0];  // successful

    ReadOnlySpan<string> span2 = new string[] { "1" };
    string temp2 = span2[0];    // fails with IndexOutOfRangeException
}
```
```xml
<test name="System.SpanTests.ReadOnlySpanTests.TestingSpanReferenceTypeForUAPAOT" type="System.SpanTests.ReadOnlySpanTests" method="TestingSpanReferenceTypeForUAPAOT" time="0.0000441" result="Fail">
    <failure exception-type="System.IndexOutOfRangeException">
        <message><![CDATA[System.IndexOutOfRangeException : Index was outside the bounds of the array.]]></message>
        <stack-trace><![CDATA[   at Internal.Runtime.CompilerHelpers.ThrowHelpers.ThrowIndexOutOfRangeException() in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\CompilerHelpers\ThrowHelpers.cs:line 25
            at System.SpanTests.ReadOnlySpanTests.TestingSpanReferenceTypeForUAPAOT(ReadOnlySpan$1<Byte> span1) in D:\GitHub\Fork\corefx\src\System.Memory\tests\ReadOnlySpan\CtorArray.cs:line 66
            at _$ILCT$.$ILT$ReflectionDynamicInvoke$.InvokeRetV(Object thisPtr, IntPtr methodToCall, ArgSetupState argSetupState, Boolean targetIsThisCall)
            at System.InvokeUtils.CalliIntrinsics.Call(IntPtr dynamicInvokeHelperMethod, Object thisPtr, IntPtr methodToCall, ArgSetupState argSetupState, Boolean isTargetThisCall)
            at System.InvokeUtils.CallDynamicInvokeMethod(Object thisPtr, IntPtr methodToCall, Object thisPtrDynamicInvokeMethod, IntPtr dynamicInvokeHelperMethod, IntPtr dynamicInvokeHelperGenericDictionary, Object targetMethodOrDelegate, Object[] parameters, BinderBundle binderBundle, Boolean wrapInTargetInvocationException, Boolean invokeMethodHelperIsThisCall, Boolean methodToCallIsThisCall, Boolean parametersNeedCopyBack, ArgSetupState argSetupState, Object[] parametersOld, Object[] nullableCopyBackObjectsOld, Int32 curIndexOld, Object targetMethodOrDelegateOld, BinderBundle binderBundleOld, Object[] customBinderProvidedParametersOld, Object result, Exception e, Int32 i) in CallDynamicInvokeMethod:line 16707566
        ]]></stack-trace>
    </failure>
```

**The Unsafe.RefAreSame test fails on uapaot:**
```C#
public static void RefAreSame()
{
    long[] a = new long[2];
    Assert.True(Unsafe.AreSame(ref a[0], ref a[0])); // successful
    Assert.False(Unsafe.AreSame(ref a[0], ref a[1])); // fails
}
```

```xml
<test name="System.Runtime.CompilerServices.UnsafeTests.RefAreSame" type="System.Runtime.CompilerServices.UnsafeTests" method="RefAreSame" time="0.0003107" result="Fail">
    <failure exception-type="Xunit.Sdk.FalseException">
        <message><![CDATA[Assert.False() Failure\r\nExpected: False\r\nActual:   True]]></message>
        <stack-trace><![CDATA[   at System.Runtime.CompilerServices.UnsafeTests.RefAreSame(Int64[] a) in D:\GitHub\Fork\corefx\src\System.Runtime.CompilerServices.Unsafe\tests\UnsafeTests.cs:line 593
            at _$ILCT$.$ILT$ReflectionDynamicInvoke$.InvokeRetV(Object thisPtr, IntPtr methodToCall, ArgSetupState argSetupState, Boolean targetIsThisCall)
            at System.InvokeUtils.CalliIntrinsics.Call(IntPtr dynamicInvokeHelperMethod, Object thisPtr, IntPtr methodToCall, ArgSetupState argSetupState, Boolean isTargetThisCall)
            at System.InvokeUtils.CallDynamicInvokeMethod(Object thisPtr, IntPtr methodToCall, Object thisPtrDynamicInvokeMethod, IntPtr dynamicInvokeHelperMethod, IntPtr dynamicInvokeHelperGenericDictionary, Object targetMethodOrDelegate, Object[] parameters, BinderBundle binderBundle, Boolean wrapInTargetInvocationException, Boolean invokeMethodHelperIsThisCall, Boolean methodToCallIsThisCall, Boolean parametersNeedCopyBack, ArgSetupState argSetupState, Object[] parametersOld, Object[] nullableCopyBackObjectsOld, Int32 curIndexOld, Object targetMethodOrDelegateOld, BinderBundle binderBundleOld, Object[] customBinderProvidedParametersOld, Object result, Exception e, Int32 i) in CallDynamicInvokeMethod:line 16707566
        ]]></stack-trace>
    </failure>
```

I am not sure why Span\<reference type\> indexer (`ref T this[int index]`) and Unsafe.AreSame\<T\>(ref T left, ref T right) are failing on uapaot. Any ideas?

cc @jkotas, @safern, @danmosemsft

### CODE REVIEW

**Comment by jkotas** [comment]
>  failing on uapaot. Any ideas?

They are uapaot bugs that need to be fixed. Open bugs on these in TFS and assign them to ianb.

**Comment by safern** [comment]
> They are uapaot bugs that need to be fixed. Open bugs on these in TFS and assign them to ianb.

Also disable the tests with `[ActiveIssue("bug link", TargetFrameworkMonikers.UapAot]` note that we rather track the bug links in active issue to be in GitHub, so you could open an issue in GitHub to track that disabled test and just link the TFS bug to that issue.

**Comment by karelz** [comment]
Do we need this PR also in UWP6.0 branch?

**Comment by jkotas** [comment]
> Do we need this PR also in UWP6.0 branch?

I do not think so. Span is not shipping in UWP6.0.

File: src/System.Memory/tests/Memory/Span.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ public static void SpanFromCtorArrayLong()
             owner.AsMemory.Span.Validate(91, -92, 93, 94, -95);
         }
 
+        [ActiveIssue(20884, TargetFrameworkMonikers.UapAot)]
```
Comment (safern) [inline src/System.Memory/tests/Memory/Span.cs]: Why are you linking it to a System.Drawing issue? https://github.com/dotnet/corefx/issues/20884

File: src/System.Memory/tests/ReadOnlyMemory/CtorArray.cs [hunk 2]
```diff
@@ -41,6 +41,7 @@ public static void CtorArrayLong()
             memory.Validate(91, -92, 93, 94, -95);
         }
 
+        [ActiveIssue(20884, TargetFrameworkMonikers.UapAot)]
```
Comment (safern) [inline src/System.Memory/tests/ReadOnlyMemory/CtorArray.cs]: Why are you linking it to a System.Drawing issue? #20884

File: src/System.Memory/tests/ReadOnlyMemory/CtorArray.cs [hunk 3]
```diff
@@ -92,6 +93,7 @@ public static void CtorArrayWrongValueType()
             memory.Validate(42, -1);
         }
 
+        [ActiveIssue(20884, TargetFrameworkMonikers.UapAot)]
```
Comment (safern) [inline src/System.Memory/tests/ReadOnlyMemory/CtorArray.cs]: Why are you linking it to a System.Drawing issue? #20884

File: src/System.Memory/tests/ReadOnlySpan/CtorArray.cs [hunk 4]
```diff
@@ -41,6 +41,7 @@ public static void CtorArray2()
             span.Validate(91, -92, 93, 94, -95);
         }
 
+        [ActiveIssue(20884, TargetFrameworkMonikers.UapAot)]
```
Comment (safern) [inline src/System.Memory/tests/ReadOnlySpan/CtorArray.cs]: Ditto.

File: src/System.Memory/tests/ReadOnlySpan/CtorArray.cs [hunk 5]
```diff
@@ -92,6 +93,7 @@ public static void CtorArrayWrongValueType()
             span.Validate(42, -1);
         }
 
+        [ActiveIssue(20884, TargetFrameworkMonikers.UapAot)]
```
Comment (safern) [inline src/System.Memory/tests/ReadOnlySpan/CtorArray.cs]: Ditto.

File: src/System.Memory/tests/Span/AsSpan.cs [hunk 6]
```diff
@@ -35,18 +44,27 @@ public static void ZeroLengthArrayAsSpan()
...
+        [Fact]
+        public static void LongArraySegmentAsSpan()
+        {
             long[] b = { 91, -92, 93, 94, -95 };
             ArraySegment<long> segmentLong = new ArraySegment<long>(b, 1, 3);
             Span<long> spanLong = segmentLong.AsSpan();
             spanLong.Validate(-92, 93, 94);
+        }
 
+        [ActiveIssue(20884, TargetFrameworkMonikers.UapAot)]
```
Comment (safern) [inline src/System.Memory/tests/Span/AsSpan.cs]: Ditto.

File: src/System.Memory/tests/Span/CtorArray.cs [hunk 7]
```diff
@@ -40,6 +40,7 @@ public static void CtorArray2()
             span.Validate(91, -92, 93, 94, -95);
         }
 
+        [ActiveIssue(20884, TargetFrameworkMonikers.UapAot)]
```
Comment (safern) [inline src/System.Memory/tests/Span/CtorArray.cs]: Ditto.

File: src/System.Memory/tests/Memory/Span.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/Memory/Span.cs]: Oops. Forgot to hit save :)

File: src/System.Memory/tests/Span/CtorArray.cs (hunk 7 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/Span/CtorArray.cs]: Fixed.

File: src/System.Memory/tests/ReadOnlyMemory/CtorArray.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlyMemory/CtorArray.cs]: Fixed. Thanks for catching it.

File: src/System.Memory/tests/Memory/Span.cs (hunk 1 above)
Comment (safern) [reply src/System.Memory/tests/Memory/Span.cs]: :)

**Pull request merged**